src/Microsoft.AspNetCore.Mvc.Api.Analyzers/ActualApiResponseMetadataFactory.cs

```diff
@@ -0,0 +1,292 @@
+                return null;
+            }
+
+            var statementReturnType = typeInfo.Type;
+
+            if (!symbolCache.IActionResult.IsAssignableFrom(statementReturnType))
+            {
+                // Return expression is not an instance of IActionResult. Must be returning the "model".
+                return new ActualApiResponseMetadata(returnStatementSyntax, statementReturnType);
+            }
+
+            var defaultStatusCodeAttribute = statementReturnType
+                .GetAttributes(symbolCache.DefaultStatusCodeAttribute, inherit: true)
+                .FirstOrDefault();
+
+            var statusCode = GetDefaultStatusCode(defaultStatusCodeAttribute);
+            ITypeSymbol returnType = null;
+            switch (returnExpression)
+            {
+                case InvocationExpressionSyntax invocation:
+                    {
+                        // Covers the 'return StatusCode(200)' case.
+                        var result = InspectMethodArguments(symbolCache, semanticModel, invocation.Expression, invocation.ArgumentList, cancellationToken);
+                        statusCode = result.statusCode ?? statusCode;
+                        returnType = result.returnType;
+                        break;
+                    }
+
+                case ObjectCreationExpressionSyntax creation:
+                    {
+                        // Recover values from 'return new StatusCodeResult(200) case.
+                        var result = InspectMethodArguments(symbolCache, semanticModel, creation, creation.ArgumentList, cancellationToken);
+                        statusCode = result.statusCode ?? statusCode;
+                        returnType = result.returnType;
+
+                        // Recover values from property assignments e.g. 'return new ObjectResult(...) { StatusCode = 200 }'.
+                        // Property assignments override constructor assigned values and defaults.
+                        result = InspectInitializers(symbolCache, semanticModel, creation.Initializer, cancellationToken);
+                        statusCode = result.statusCode ?? statusCode;
+                        returnType = result.returnType ?? returnType;
+                        break;
+                    }
+            }
+
+            if (statusCode == null)
+            {
+                return null;
+            }
+
+            return new ActualApiResponseMetadata(returnStatementSyntax, statusCode.Value, returnType);
+        }
+
+        private static (int? statusCode, ITypeSymbol returnType) InspectInitializers(
+            in ApiControllerSymbolCache symbolCache,
+            SemanticModel semanticModel,
+            InitializerExpressionSyntax initializer,
+            CancellationToken cancellationToken)
+        {
+            int? statusCode = null;
+            ITypeSymbol typeSymbol = null;
+
+            for (var i = 0; initializer != null && i < initializer?.Expressions.Count; i++)
```

Comment:
This null check isn't really needed btw, `i < null` will return false for all values of `i`